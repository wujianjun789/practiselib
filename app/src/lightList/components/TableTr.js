/**
 * Created by Azrael on 2017/9/15.
 */
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
/**
 * 这里是注释
 * 这里是注释
 * 这里是注释
 * 这里是注释
 */
export default class TableTr extends PureComponent {
    static propTypes = {
        activeId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        columns: PropTypes.arrayOf(PropTypes.shape({
            accessor: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
            title: PropTypes.string.isRequired
        }) ),
        willMountFuncs: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.func),
            PropTypes.func
        ]),
        rowClick: PropTypes.func
    }

    constructor(props) {
        super(props);
        this.state = {
            data: {...props.data},
        };

        this.onClick = this.onClick.bind(this);
        this.updateData = this.updateData.bind(this);
    }

    componentWillMount() {
        this.mounted = true;
        const {willMountFuncs} = this.props;
        if (Array.isArray(willMountFuncs)) {
            willMountFuncs.forEach(func => {func(this.updateData)});
        } else if (typeof willMountFuncs == 'function') {
            willMountFuncs(this.updateData);
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    updateData(data) {
        this.mounted && this.setState({data: {...this.state.data, ...data}});
    }

    onClick() {
        this.props.rowClick && this.props.rowClick(this.state.data);
    }

    render() {
        const {data, data:{id}} = this.state;
        const {activeId, columns} = this.props;

        return (
            <tr className={(activeId || (typeof activeId == 'number' && activeId == 0) ) && activeId==id ? 'active':''} onClick={this.onClick}>
            {
                columns.map((item, index) => {
                    if(typeof item.accessor === 'function') {
                        return <td key={index}>{item.accessor(data)}</td>
                    } else if(typeof item.accessor === 'string') {
                        return <td key={index}>{data[item.accessor]}</td>
                    } else {
                        return <td key={index}></td>
                    }
                })
            }
            </tr>
        )
    }
}
