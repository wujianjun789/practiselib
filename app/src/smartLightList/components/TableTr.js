/**
 * Created by Azrael on 2017/9/15.
 */
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
/**
 * 这里是注释
 * 这里是注释
 * 这里是注释
 * 这里是注释
 */
export default class TableTr extends PureComponent {
    static propTypes = {
        activeId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        columns: PropTypes.arrayOf(PropTypes.shape({
            accessor: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
            title: PropTypes.string.isRequired
        }) ),
        formatFunc: PropTypes.func,
        willMountFuncs: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.func),
            PropTypes.func
        ]),
        rowClick: PropTypes.func
    }

    constructor(props) {
        super(props);
        this.state = {
            data: {...props.data},
            activeId: props.activeId
        };

        this.onClick = this.onClick.bind(this);
        this.updateData = this.updateData.bind(this);
    }

    componentWillMount() {
        this.mounted = true;
        const {willMountFuncs} = this.props;
        if (Array.isArray(willMountFuncs)) {
            willMountFuncs.forEach(func => {func(this.updateData)});
        } else if (typeof willMountFuncs == 'function') {
            willMountFuncs(this.updateData);
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    updateData(data) {
        this.mounted && this.setState({data: {...this.state.data, ...data}});
    }

    onClick() {
        this.props.rowClick && this.props.rowClick(this.state.data);
    }

    render() {
        const {data, data:{id}} = this.state;
        const {activeId, columns, formatFunc} = this.props;

        let _data = {...data};
        if(typeof formatFunc == 'function') {
            formatFunc(_data);
         }

        return (
            <tr className={activeId && activeId==id ? 'active':''} onClick={this.onClick}>
            {
                columns.map((item, index) => {
                    if(typeof item.accessor === 'function') {
                        return <td key={index}>{item.accessor(_data)}</td>
                    } else if(typeof item.accessor === 'string') {
                        return <td key={index}>{_data[item.accessor]}</td>
                    } else {
                        return <td key={index}></td>
                    }
                })
            }
            </tr>
        )
    }
}