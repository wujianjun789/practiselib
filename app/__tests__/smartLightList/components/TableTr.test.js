import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import TableTr from '../../../src/smartLightList/components/TableTr';

describe('', ()=> {
	const columns = [
		{accessor: 'name', title: '设备名称'},
		{accessor(data){return data.value*10}, title: 'value*10'}
	];
	const data = {id: 0, name: 'test', value: 10};

	it('default render', () => {
		const cmp = shallow(<TableTr data={data} columns={columns}/>);
		const tr = cmp.find('tr');
		expect(!tr.hasClass('active')).toBeTruthy();
		const td = tr.find('td');
		expect(td.length).toBe(columns.length);
		columns.forEach((item, index) => {
			expect(td.at(index).text()).toBe(typeof item.accessor === 'function'?''+item.accessor(data):''+data[item.accessor]);
		})
	});

	it('render with activeId', () => {
		const cmp = shallow(<TableTr data={data} columns={columns} activeId={data.id}/>);
		expect(cmp.find('tr').hasClass('active')).toBeTruthy();
	});

	it('simulate click', () => {
		const fn = jest.fn();
		const cmp = shallow(<TableTr data={data} columns={columns} activeId={data.id} rowClick={fn} />);
		cmp.simulate('click', data);
		expect(fn).toHaveBeenCalledTimes(1);
		expect(fn).toHaveBeenCalledWith(data);
	});

	it('default render snapshot', () => {
		const cmp = renderer.create(<TableTr data={data} columns={columns}/>);
		expect(cmp.toJSON()).toMatchSnapshot();
	});

	it('render with activeId snapshot', () => {
		const cmp = renderer.create(<TableTr data={data} columns={columns} activeId={data.id}/>);
		expect(cmp.toJSON()).toMatchSnapshot();
	});

	it('simulate click snapshot', () => {
		const fn = jest.fn();
		const cmp = renderer.create(<TableTr data={data} columns={columns} activeId={data.id} rowClick={fn} />);
		expect(cmp.toJSON()).toMatchSnapshot();
	});
});
