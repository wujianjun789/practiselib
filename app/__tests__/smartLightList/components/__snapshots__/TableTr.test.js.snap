exports[` default render snapshot 1`] = `
<tr
  className=""
  onClick={[Function]}>
  <td>
    test
  </td>
  <td>
    100
  </td>
</tr>
`;

exports[` render with activeId snapshot 1`] = `
<tr
  className="active"
  onClick={[Function]}>
  <td>
    test
  </td>
  <td>
    100
  </td>
</tr>
`;

exports[` simulate click snapshot 1`] = `
<tr
  className="active"
  onClick={[Function]}>
  <td>
    test
  </td>
  <td>
    100
  </td>
</tr>
`;
